src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1233,33 +1256,112 @@ private Task CopyToUntilEofAsync(Stream destination, int bufferSize, Cancellatio
+
+            // If the last request already completed (because the response had no content), return the connection to the pool now.
+            // Otherwise, it will be returned when the response has been consumed and CompleteResponse below is called.
+            if (_currentRequest == null)
             {
-                Trace("Server requested connection be closed.");
+                ReturnConnectionToPool();
             }
+        }
+
+        private void CompleteResponse()
+        {
+            Debug.Assert(_currentRequest != null, "Expected the connection to be associated with a request.");
+            Debug.Assert(_writeOffset == 0, "Everything in write buffer should have been flushed.");
+
+            // Disassociate the connection from a request.
+            _currentRequest = null;
 
             // If we have extraneous data in the read buffer, don't reuse the connection;
             // otherwise we'd interpret this as part of the next response.
-            if (_readOffset != _readLength)
+            if (RemainingBuffer.Length != 0)
             {
+                if (NetEventSource.IsEnabled)
+                {
+                    Trace("Unexpected data on connection after response read.");
+                }
+
+                ConsumeFromRemainingBuffer(RemainingBuffer.Length);
                 _connectionClose = true;
             }
 
-            // If server told us it's closing the connection, don't put this back in the pool.
-            // And if we incurred an error while transferring request content, also skip the pool.
-            if (!_connectionClose)
+            // If the connection is no longer in use (i.e. for NT authentication), then we can return it to the pool now.
+            // Otherwise, it will be returned when the connection is no longer in use (i.e. Release above is called).
+            if (!_inUse)
             {
-                try
+                ReturnConnectionToPool();
+            }
+        }
+
+        public async Task DrainResponseAsync(HttpResponseMessage response)
+        {
+            Debug.Assert(_inUse);
+
+            if (_connectionClose)
+            {
+                throw new HttpRequestException(SR.net_http_authconnectionfailure);
+            }
+
+            HttpContentReadStream responseStream = (HttpContentReadStream)await response.Content.ReadAsStreamAsync();
+
+            if (responseStream.NeedsDrain)
+            {
+                Debug.Assert(response.RequestMessage == _currentRequest);
+
+                if (!await responseStream.DrainAsync(_pool.Settings._maxResponseDrainSize))
+                {
+                    throw new HttpRequestException(SR.net_http_authconnectionfailure);
+                }
+
+                // Draining may have set _connectionClose, so check again
+                if (_connectionClose)
+                {
+                    throw new HttpRequestException(SR.net_http_authconnectionfailure);
+                }
```

Comment:
Can we combine these to simplify / reduce the number of throw sites? e.g.
```C#
if (!await responseStream.DrainAsync(_pool.Settings._maxResponseDrainSize).ConfigureAwait(false) ||
    _connectionClose) // draining may have set _connectionClose, so check again
{
    throw new HttpRequestException(SR.net_http_authconnectionfailure);
}
```

Reply:
Sure